spidy/spidy.go

```diff
@@ -0,0 +1,381 @@
+package spidy
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/ardanlabs/kit/pool"
+)
+
+// Events defines an interface to allows the logging of events as they occur
+// within the spidy API.
+type Events interface {
+	Event(context interface{}, event string, format string, data ...interface{})
+	ErrorEvent(context interface{}, event string, err error, format string, data ...interface{})
+}
+
+//==============================================================================
+
+// LinkReport defines a struct to entail failed links with their status and errors.
+type LinkReport struct {
+	Link   string
+	Status int
+	Error  error
+}
+
+// Run evaluates the given urlPath returning possible lists of deadlinks found
+// within the page of the given link else returns a non-nil error if it failed.
+func Run(context interface{}, urlPath string, all bool, workers int, depths int, events Events) ([]LinkReport, error) {
+	events.Event(context, "Run", "Started : URL[%s] : Include Externals[%t] : Workers[%d]", urlPath, all, workers)
+
+	path, err := url.Parse(urlPath)
+	if err != nil {
+		events.ErrorEvent(context, "Run", err, "Completed")
+		return nil, err
+	}
+
+	// deadLinks provides a lists to contain all failed url links that are
+	// considered deadlinks.
+	var deadLinks []LinkReport
+
+	dead := make(chan LinkReport)
+
+	go collectFrom(path, all, workers, depths, dead, events)
+
+	for link := range dead {
+		deadLinks = append(deadLinks, link)
+	}
+
+	events.Event(context, "Run", "Completed : Total Dead Links[%d]", len(deadLinks))
+	return deadLinks, nil
+}
+
+//==============================================================================
+
+var depths int64
+
+// collectFrom uses a recursive function to map out the needed lists of links to.
+// It returns a channel through which the acceptable links can be crawled from.
+func collectFrom(path *url.URL, doExternals bool, maxWorkers int, depths int, dead chan LinkReport, events Events) {
+	poolCfg := pool.Config{
+		OptEvent:    pool.OptEvent{Event: events.Event},
+		MinRoutines: func() int { return 10 },
+		MaxRoutines: func() int { return maxWorkers },
+	}
+
+	// create a new pool.
+	pl, err := pool.New("spidy", "collectFrom", poolCfg)
+	if err != nil {
+		fmt.Printf("Spidy failed to create work pool: %s\n", err.Error())
+		close(dead)
+		return
+	}
+
+	status, crawleable, err := evaluatePath(path.String())
+	if err != nil {
+		pl.Shutdown("spidy")
```

Comment:
There is no need for these statements if you move the defer statements on line 89 & 90 above the evaluatePath function call